### ISSUE TITLE
Question: why not just use named pipe on Windows and unix domain socket on Linux/Mac?

### ISSUE THREAD

**Issue by asashnov** [itay-grudev/SingleApplication#45]
Question: why not just use named pipe on Windows and unix domain socket on Linux/Mac?

Maybe it is possible to implement the same functionality with just named pipe, without shared memory segment?
The main application tries to create named pipe or unix domain socket like /tmp/single-app-my-application.socket    If it's already created- try to connect.

**Comment by itay-grudev** [comment]
When I did my experiments several years ago, I concluded that a shared memory segment with a semaphore (implemented in the `QSharedMemory`) is the most reliable solution.

**Comment by x1ddos** [comment]
https://doc.qt.io/qt-5/qlocalserver.html#listen docs say:

> On Windows two local servers can listen to the same pipe at the same time, but any connections will go to one of the server.

@itay-grudev do you remember from your notes if that could be one of the reasons to use shared memory as a semaphore? It sounds like simply `listen`'ing on a pipe may lead to some race conditions, although it seems to kind of contradict an earlier paragraph:

> If the server is currently listening then it will return false. Return true on success otherwise false.

@benma and I are looking into a "single qt app instance" approach and this lib came up as a nice solution. Still, it would be very good to know the reasoning behind the design.

**Comment by itay-grudev** [comment]
Local servers are unreliable in terms of race conditions, while shared memory and shared locks work consistently across platforms. QSharedMemory includes a shared semaphore in it's implementation, although I did initially experiment with a another semaphore inside it, although it wasn't necessary.

So I ended up using shared memory with a lock for exclusivity and identifying a primary instances, while local servers for inter-instance communication.

There used to be an issue on *nix systems, where the shared memory remains if the lost process doesn't close it properly, but I discovered that attaching to it and then deallocating it clears it. So the current implementation does one attach and deallocation prior to using it, just to make sure there isn't a crashed primary instance.
